File: app/schemas/account.py
import re
from typing import Literal

from pydantic import BaseModel, EmailStr, field_validator


class UserSignUp(BaseModel):
    email: EmailStr
    password: str
    display_name: str | None = None

    @field_validator("password")
    @classmethod
    def password_strength(cls, v: str) -> str:
        if len(v) < 8:
            raise ValueError("Password must be at least 8 characters.")
        if not re.search(r"[A-Za-z]", v):
            raise ValueError("Password must contain at least one letter.")
        if not re.search(r"[0-9!@#$%^&*()_+\-=\[\]{};':\"\\|,.<>/?]", v):
            raise ValueError("Password must contain at least one number or special character.")
        return v


class UserSignIn(BaseModel):
    email: str
    password: str


class RefreshSessionRequest(BaseModel):
    refresh_token: str


class AuthResponse(BaseModel):
    access_token: str
    refresh_token: str
    user_id: str
    email: str


class UserProfile(BaseModel):
    id: str
    email: str
    display_name: str | None = None
    avatar_url: str | None = None
    preferred_translation: str = "WEB"
    preferred_language: str = "en"


class UserProfileUpdate(BaseModel):
    display_name: str | None = None
    avatar_url: str | None = None
    preferred_translation: str | None = None
    preferred_language: str | None = None


StudyItemKind = Literal["highlight", "bookmark", "note"]


class StudyItemCreate(BaseModel):
    translation: str
    book_number: int
    chapter: int
    verse_start: int
    verse_end: int | None = None
    kind: StudyItemKind
    color: str | None = None
    note: str | None = None


class StudyItemUpdate(BaseModel):
    translation: str | None = None
    book_number: int | None = None
    chapter: int | None = None
    verse_start: int | None = None
    verse_end: int | None = None
    kind: StudyItemKind | None = None
    color: str | None = None
    note: str | None = None


class StudyItemOut(BaseModel):
    id: int
    user_id: str
    translation: str
    book_number: int
    chapter: int
    verse_start: int
    verse_end: int | None = None
    kind: StudyItemKind
    color: str | None = None
    note: str | None = None
    created_at: str | None = None
    updated_at: str | None = None


class ReadingProgressIn(BaseModel):
    translation: str
    book_number: int
    chapter: int
    last_verse: int = 1


class ReadingProgressOut(BaseModel):
    id: int | None = None
    user_id: str
    translation: str
    book_number: int
    chapter: int
    last_verse: int = 1
    updated_at: str | None = None


DefaultPanel = Literal["none", "commentary", "ai", "study"]
FontScale = Literal["8", "10", "12", "14", "16", "18", "20", "22", "24", "26", "28", "30"]
LineHeight = Literal["1.5", "2", "2.5"]
SidePanelPosition = Literal["left", "right"]
ReaderFont = Literal["georgia", "charter", "palatino", "garamond", "times", "sans", "openSans", "dyslexie"]
HighlightColor = Literal["yellow", "amber", "green", "blue", "pink", "lavender", "mint"]
ReadingMode = Literal["single", "book", "paged-single", "paged-double"]


class RecentPassage(BaseModel):
    book: str
    chapter: int
    translation: str
    saved_at: int


class UserSettingsIn(BaseModel):
    theme: str = "default"
    font_scale: FontScale = "20"
    line_height: LineHeight = "2"
    reduced_motion: bool = False
    page_flip_enabled: bool = True
    default_panel: DefaultPanel = "none"
    side_panel_position: SidePanelPosition = "right"
    reader_font: ReaderFont = "georgia"
    highlight_color: HighlightColor = "yellow"
    reading_mode: ReadingMode = "single"
    recent_passages: list[RecentPassage] = []


class UserSettingsOut(UserSettingsIn):
    user_id: str
    updated_at: str | None = None

File: frontend-next/src/lib/account.ts
export const AUTH_SESSION_STORAGE_KEY = 'bible-app-auth-session';

export interface AuthSession {
  accessToken: string;
  refreshToken: string;
  userId: string;
  email: string;
}

export interface AccountProfile {
  id: string;
  email: string;
  display_name: string | null;
  avatar_url: string | null;
  preferred_translation: string;
  preferred_language: string;
}

export interface SyncedStudyItem {
  id?: number;
  user_id?: string;
  translation: string;
  book_number: number;
  chapter: number;
  verse_start: number;
  verse_end?: number | null;
  kind: 'highlight' | 'bookmark' | 'note';
  color?: string | null;
  note?: string | null;
}

export interface SyncedReadingProgress {
  translation: string;
  book_number: number;
  chapter: number;
  last_verse: number;
}

export interface SyncedRecentPassage {
  book: string;
  chapter: number;
  translation: string;
  saved_at: number;
}

export type SyncedFontScale =
  | '8'
  | '10'
  | '12'
  | '14'
  | '16'
  | '18'
  | '20'
  | '22'
  | '24'
  | '26'
  | '28'
  | '30';
export type SyncedLineHeight = '1.5' | '2' | '2.5';
export type SyncedReaderFont =
  | 'georgia'
  | 'charter'
  | 'palatino'
  | 'garamond'
  | 'times'
  | 'sans'
  | 'openSans'
  | 'dyslexie';
export type SyncedHighlightColor =
  | 'yellow'
  | 'amber'
  | 'green'
  | 'blue'
  | 'pink'
  | 'lavender'
  | 'mint';
export type SyncedReadingMode =
  | 'single'
  | 'book'
  | 'paged-single'
  | 'paged-double';

export interface SyncedUserSettings {
  user_id?: string;
  theme: string;
  font_scale: SyncedFontScale;
  line_height: SyncedLineHeight;
  reduced_motion: boolean;
  page_flip_enabled: boolean;
  default_panel: 'none' | 'commentary' | 'ai' | 'study';
  side_panel_position: 'left' | 'right';
  /** Absent on older API responses; client defaults to Georgia / yellow. */
  reader_font?: SyncedReaderFont;
  highlight_color?: SyncedHighlightColor;
  reading_mode?: SyncedReadingMode;
  recent_passages: SyncedRecentPassage[];
}

interface RawAuthResponse {
  access_token: string;
  refresh_token: string;
  user_id: string;
  email: string;
}

function toSession(payload: RawAuthResponse): AuthSession {
  return {
    accessToken: payload.access_token,
    refreshToken: payload.refresh_token,
    userId: payload.user_id,
    email: payload.email,
  };
}

function formatApiError(payload: unknown, status: number): string {
  if (payload && typeof payload === 'object' && 'detail' in payload) {
    const detail = (payload as { detail?: unknown }).detail;
    if (typeof detail === 'string') return detail;
    if (Array.isArray(detail)) {
      const parts = detail.map((item: unknown) => {
        if (item && typeof item === 'object' && 'msg' in item) {
          return String((item as { msg: unknown }).msg);
        }
        return typeof item === 'string' ? item : JSON.stringify(item);
      });
      return parts.filter(Boolean).join(' ') || `Request failed: ${status}`;
    }
  }
  return `Request failed: ${status}`;
}

async function request<T>(
  path: string,
  init?: RequestInit,
  accessToken?: string,
): Promise<T> {
  const headers = new Headers(init?.headers ?? {});
  if (!headers.has('Content-Type') && init?.body) headers.set('Content-Type', 'application/json');
  if (accessToken) headers.set('Authorization', `Bearer ${accessToken}`);
  const response = await fetch(path, { ...init, headers });
  if (!response.ok) {
    const payload = await response.json().catch(() => null);
    throw new Error(formatApiError(payload, response.status));
  }
  if (response.status === 204) return undefined as T;
  return response.json() as Promise<T>;
}

function parseStoredSession(raw: string | null): AuthSession | null {
  if (!raw) return null;
  try {
    const parsed = JSON.parse(raw) as AuthSession;
    return parsed?.accessToken && parsed?.refreshToken ? parsed : null;
  } catch {
    return null;
  }
}

/** Restores session from localStorage (remember-me / legacy) or sessionStorage (this browser session only). */
export function restoreAuthSession(): AuthSession | null {
  if (typeof window === 'undefined') return null;
  return (
    parseStoredSession(window.localStorage.getItem(AUTH_SESSION_STORAGE_KEY))
    ?? parseStoredSession(window.sessionStorage.getItem(AUTH_SESSION_STORAGE_KEY))
  );
}

/**
 * Persists auth tokens. Use `remember: true` for localStorage (survives browser restarts),
 * `remember: false` for sessionStorage (cleared when the tab/window session ends).
 * Omit `remember` when refreshing tokens to keep the same storage as the active session.
 */
export function persistAuthSession(session: AuthSession | null, remember?: boolean) {
  if (typeof window === 'undefined') return;
  if (!session) {
    window.localStorage.removeItem(AUTH_SESSION_STORAGE_KEY);
    window.sessionStorage.removeItem(AUTH_SESSION_STORAGE_KEY);
    return;
  }
  const payload = JSON.stringify(session);
  let useLocal: boolean;
  if (remember === true) useLocal = true;
  else if (remember === false) useLocal = false;
  else {
    if (window.localStorage.getItem(AUTH_SESSION_STORAGE_KEY)) useLocal = true;
    else if (window.sessionStorage.getItem(AUTH_SESSION_STORAGE_KEY)) useLocal = false;
    else useLocal = true;
  }
  if (useLocal) {
    window.localStorage.setItem(AUTH_SESSION_STORAGE_KEY, payload);
    window.sessionStorage.removeItem(AUTH_SESSION_STORAGE_KEY);
  } else {
    window.sessionStorage.setItem(AUTH_SESSION_STORAGE_KEY, payload);
    window.localStorage.removeItem(AUTH_SESSION_STORAGE_KEY);
  }
}

export async function signUp(email: string, password: string, displayName?: string) {
  const response = await request<RawAuthResponse>('/api/v1/auth/signup', {
    method: 'POST',
    body: JSON.stringify({ email, password, display_name: displayName || null }),
  });
  return toSession(response);
}

export async function signIn(email: string, password: string) {
  const response = await request<RawAuthResponse>('/api/v1/auth/signin', {
    method: 'POST',
    body: JSON.stringify({ email, password }),
  });
  return toSession(response);
}

export async function refreshSession(session: AuthSession) {
  const response = await request<RawAuthResponse>('/api/v1/auth/refresh', {
    method: 'POST',
    body: JSON.stringify({ refresh_token: session.refreshToken }),
  });
  return toSession(response);
}

export async function signOut(session: AuthSession) {
  await request<void>('/api/v1/auth/signout', { method: 'POST' }, session.accessToken);
}

export async function getProfile(session: AuthSession) {
  return request<AccountProfile>('/api/v1/me', undefined, session.accessToken);
}

export async function updateProfile(session: AuthSession, payload: Partial<AccountProfile>) {
  return request<AccountProfile>('/api/v1/me', {
    method: 'PATCH',
    body: JSON.stringify(payload),
  }, session.accessToken);
}

export async function getStudyItems(session: AuthSession) {
  return request<SyncedStudyItem[]>('/api/v1/study-items', undefined, session.accessToken);
}

export async function putStudyItems(session: AuthSession, payload: SyncedStudyItem[]) {
  return request<SyncedStudyItem[]>('/api/v1/study-items', {
    method: 'PUT',
    body: JSON.stringify(payload),
  }, session.accessToken);
}

export async function getReadingProgress(session: AuthSession) {
  return request<SyncedReadingProgress[]>('/api/v1/reading-progress', undefined, session.accessToken);
}

export async function putReadingProgress(session: AuthSession, payload: SyncedReadingProgress[]) {
  return request<SyncedReadingProgress[]>('/api/v1/reading-progress', {
    method: 'PUT',
    body: JSON.stringify(payload),
  }, session.accessToken);
}

export async function getUserSettings(session: AuthSession) {
  return request<SyncedUserSettings>('/api/v1/user-settings', undefined, session.accessToken);
}

export async function putUserSettings(session: AuthSession, payload: SyncedUserSettings) {
  return request<SyncedUserSettings>('/api/v1/user-settings', {
    method: 'PUT',
    body: JSON.stringify(payload),
  }, session.accessToken);
}
